src/Microsoft.AspNet.Proxy/ProxyMiddleware.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Proxy
+{
+    public class ProxyMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly HttpClient _httpClient;
+        private readonly ProxyOptions _options;
+
+        public ProxyMiddleware([NotNull] RequestDelegate next, [NotNull] ProxyOptions options)
+        {
+            _next = next;
+            if (string.IsNullOrEmpty(options.Host))
+            {
+                throw new InvalidOperationException("Options parameter must specify host.");
+            }
+
+            // Setting default Port and Scheme if not specified
+            if (string.IsNullOrEmpty(options.Port))
+            {
+                options.Port = "80";
+            }
+
+            if (string.IsNullOrEmpty(options.Scheme))
+            {
+                options.Scheme = "http";
+            }
+
+            _options = options;
+
+#if DNX451
+            _httpClient = new HttpClient(_options.BackChannelMessageHandler?? new HttpClientHandler());
+#else
+            _httpClient = new HttpClient(_options.BackChannelMessageHandler?? new Net.Http.Client.ManagedHandler());
+#endif
+
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var requestMessage = new HttpRequestMessage();
+            if (string.Equals(context.Request.Method, "POST", StringComparison.OrdinalIgnoreCase))
+            {
+                var streamContent = new StreamContent(context.Request.Body);
+                requestMessage.Content = streamContent;
+            }
+
+            // Copy the request headers
+            foreach (var header in context.Request.Headers)
+            {
+                if (!requestMessage.Headers.TryAddWithoutValidation(header.Key, header.Value) && requestMessage.Content != null)
+                {
+                    requestMessage.Content.Headers.TryAddWithoutValidation(header.Key, header.Value);
+                }
+            }
+
+            requestMessage.Headers.Host = _options.Host + ":" + _options.Port;
+            var uriString = $"{_options.Scheme}://{_options.Host}:{_options.Port}{context.Request.PathBase}{context.Request.Path}{context.Request.QueryString}";
+            requestMessage.RequestUri = new Uri(uriString);
+            requestMessage.Method = new HttpMethod(context.Request.Method);
+            using (var responseMessage = await _httpClient.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead, context.RequestAborted))
+            {
+                context.Response.StatusCode = (int)responseMessage.StatusCode;
+                foreach (var header in responseMessage.Headers)
```

Comment:
`Location` header(ex: 201 Created)can have a absolute url which I think should be modified before sending back to the client as the client should not know the existence of the final endpoint. Probably can be handled later, file an issue?